linkerd/linkerd2
# Expand and improve edges API endpoint

This PR updates the functionality around `edges`, allowing an `--all-namespaces` flag and expanding results for `linkerd edges po -n emojivoto`.

To test, install emojivoto without linkerd:
```bash
curl -sL https://run.linkerd.io/emojivoto.yml | kubectl apply -f -
```
Then inject linkerd into only the `web` deployment:
```bash
kubectl -n emojivoto get deployments web -oyaml | bin/go-run cli inject --disable-identity - | kubectl apply -f -
```
In different terminal windows, port forward Prometheus, run the go servers and build webpack:
```bash
$ kubectl -n linkerd port-forward $(kubectl --namespace=linkerd get po --selector=linkerd.io/control-plane-component=prometheus -o jsonpath='{.items[*].metadata.name}') 9090:9090
$ go run controller/cmd/public-api/main.go --kubeconfig ~/.kube/config --prometheus-url http://localhost:9090 --log-level debug

$ cd web
$ go run main.go

$ cd web/app 
$ NODE_ENV=development yarn webpack-dev-server
```
Run CLI command:
```bash
bin/go-run cli edges deploy --all-namespaces --api-addr localhost:8085
```
You should see: 
```
SRC                  DST                  CLIENT                       SERVER                       MSG
linkerd-controller   linkerd-prometheus   linkerd-controller.linkerd   linkerd-prometheus.linkerd   -
linkerd-controller   linkerd-tap          linkerd-controller.linkerd   linkerd-tap.linkerd          -
linkerd-web          linkerd-controller   linkerd-prometheus.linkerd   linkerd-controller.linkerd   -
web                  emoji                                                                          Not Provided By Service Discovery
web                  voting                                                                         Not Provided By Service Discovery
```
Also, go to the dashboard at `localhost:8084`, you should see the resource detail page for the `web` deployment display the following:
![Screen Shot 2019-06-28 at 9 31 10 AM](https://user-images.githubusercontent.com/2289389/60357073-7ee3be00-9987-11e9-9d75-7f2c101a4dd1.png)

Summary of Changes:

API:
* If `namespace` has a value, `processEdges()` will filter Prometheus results for requests originating from or terminating in the specified namespace
* If `namespace` doesn't have a value, `processEdges()` will range through all requests
* An unsecured outbound request (a request missing a `server_id`) will be included in the edges results if there is a `resource` and a `dst_resource`

CLI:
* Adds support for `linkerd edges po --all-namespaces` (all edges between pods in all namespaces)
* Adds support for `linkerd edges po` (all edges between pods that either originate from or terminate in the default namespace)
* Expands results for `linkerd edges po -n emojivoto` (before, would return all edges between pods that both originated from or terminated into the emojivoto namespace; now, returns all edges between pods that **either originate from or terminate in** the emojivoto namespace)
* No display changes

Dashboard:
* Adds a namespace column
